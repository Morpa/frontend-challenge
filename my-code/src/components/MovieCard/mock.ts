export default [
  {
    Poster:
      'https://m.media-amazon.com/images/M/MV5BMTg2NzQwMzQyMF5BMl5BanBnXkFtZTgwNzkzODk2ODE@._V1_SX300.jpg',
    Title: 'Pele: Birth of a Legend',
    Type: 'movie',
    Year: '2016',
    imdbID: 'tt0995868'
  },
  {
    Poster:
      'https://m.media-amazon.com/images/M/MV5BMTg2NzQwMzQyMF5BMl5BanBnXkFtZTgwNzkzODk2ODE@._V1_SX300.jpg',
    Title: 'Isto É Pelé',
    Type: 'movie',
    Year: '1974',
    imdbID: 'tt0188004'
  }
]
