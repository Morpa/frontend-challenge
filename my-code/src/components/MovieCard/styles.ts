import styled, { css } from 'styled-components'
import media from 'styled-media-query'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    position: relative;
    border-radius: ${theme.border.radius};
    overflow: hidden;
    cursor: pointer;

    &:hover ${Info} {
      display: block;
    }
  `}
`

export const ImageBox = styled.div`
  height: 100%;
  width: 100%;
  background: #f6f7f8;
  background-image: linear-gradient(
    to right,
    #f6f7f8 0%,
    #edeef1 20%,
    #f6f7f8 40%,
    #f6f7f8 100%
  );
  background-size: 80rem 14rem;
  animation: placeholderShimmer 1s linear infinite forwards;

  img {
    display: block;
    width: 100%;
    height: 100%;
    object-fit: cover;
  }

  @keyframes placeholderShimmer {
    0% {
      background-position: -40rem 0;
    }

    100% {
      background-position: 40rem 0;
    }
  }
`

export const Info = styled.div`
  ${({ theme }) => css`
    display: block;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background-color: rgba(27, 35, 41, 0.8);

    > a {
      display: flex;
      justify-content: flex-end;
      flex-direction: column;
      box-sizing: border-box;
      padding: ${theme.spacings.xsmall};
      position: relative;
      height: 100%;
      align-self: flex-end;
      color: ${theme.colors.white};
      text-decoration: none;
    }
  `}

  ${media.greaterThan('medium')`
     display: none;
  `}
`
export const Title = styled.h2`
  ${({ theme }) => css`
    margin: 0;
    font-size: ${theme.font.sizes.xlarge};
    margin-bottom: ${theme.spacings.xxsmall};
  `}
`

export const Year = styled.span``

export const FavButtton = styled.div`
  ${({ theme }) => css`
    color: ${theme.colors.white};
    position: absolute;
    right: ${theme.spacings.xsmall};
    top: ${theme.spacings.xsmall};
    cursor: pointer;

    svg {
      width: 2.5rem;
    }
  `}
`
