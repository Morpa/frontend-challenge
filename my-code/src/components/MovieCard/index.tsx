import Link from 'next/link'
import { Favorite, FavoriteBorder } from '@styled-icons/material-outlined'

import * as S from './styles'

export type MovieCardProps = {
  id: string
  title: string
  poster: string
  year: string
  favorite?: boolean
  onFav?: () => void
}

const MovieCard = ({
  id,
  title,
  poster,
  year,
  favorite,
  onFav
}: MovieCardProps) => (
  <S.Wrapper>
    <S.ImageBox>
      <img src={poster} alt={`${title} movie poster`} />
    </S.ImageBox>
    <S.Info>
      <Link href={`/movie/${id}`} passHref>
        <a>
          <S.Title>{title}</S.Title>
          <S.Year>{year}</S.Year>
        </a>
      </Link>
      <S.FavButtton role="button" onClick={onFav}>
        {favorite ? (
          <Favorite aria-label="Remove from Favorites" />
        ) : (
          <FavoriteBorder aria-label="Add to Favorites" />
        )}
      </S.FavButtton>
    </S.Info>
  </S.Wrapper>
)

export default MovieCard
