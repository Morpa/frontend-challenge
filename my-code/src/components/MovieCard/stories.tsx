import { Story, Meta } from '@storybook/react/types-6-0'
import MovieCard, { MovieCardProps } from '.'

export default {
  title: 'MovieCard',
  component: MovieCard,
  args: {
    poster:
      'https://m.media-amazon.com/images/M/MV5BMTg2NzQwMzQyMF5BMl5BanBnXkFtZTgwNzkzODk2ODE@._V1_SX300.jpg',
    title: 'Pele: Birth of a Legend',
    type: 'movie',
    year: '2016',
    id: 'tt0995868'
  }
} as Meta

export const Default: Story<MovieCardProps> = (args) => (
  <div style={{ maxWidth: '30rem', margin: 'auto' }}>
    <MovieCard {...args} />
  </div>
)
