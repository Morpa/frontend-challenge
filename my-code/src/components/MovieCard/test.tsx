import { fireEvent, screen } from '@testing-library/react'

import { renderWithTheme } from 'utils/tests/helpers'

import MovieCard from '.'

const props = {
  poster:
    'https://m.media-amazon.com/images/M/MV5BMTg2NzQwMzQyMF5BMl5BanBnXkFtZTgwNzkzODk2ODE@._V1_SX300.jpg',
  title: 'Pele: Birth of a Legend',
  type: 'movie',
  year: '2016',
  id: 'tt0995868'
}

describe('<MovieCard />', () => {
  it('should render correctly', () => {
    const { container } = renderWithTheme(<MovieCard {...props} />)

    expect(
      screen.getByRole('heading', { name: props.title })
    ).toBeInTheDocument()

    expect(screen.getByText(props.year)).toBeInTheDocument()

    expect(
      screen.getByRole('img', { name: `${props.title} movie poster` })
    ).toHaveAttribute('src', props.poster)

    expect(screen.getByLabelText(/add to favorites/i)).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a filled Favorite icon when favorite is true', () => {
    renderWithTheme(<MovieCard {...props} favorite />)

    expect(screen.getByLabelText(/remove from favorites/i)).toBeInTheDocument()
  })

  it('should call onFav method when favorite is clicked', () => {
    const onFav = jest.fn()

    renderWithTheme(<MovieCard {...props} favorite onFav={onFav} />)

    fireEvent.click(screen.getAllByRole('button')[0])

    expect(onFav).toBeCalled()
  })
})
