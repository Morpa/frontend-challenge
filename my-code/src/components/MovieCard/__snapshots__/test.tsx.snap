// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieCard /> should render correctly 1`] = `
.c6 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  position: relative;
  border-radius: 0.4rem;
  overflow: hidden;
  cursor: pointer;
}

.c0:hover .c2 {
  display: block;
}

.c1 {
  height: 100%;
  width: 100%;
  background: #f6f7f8;
  background-image: linear-gradient( to right,#f6f7f8 0%,#edeef1 20%,#f6f7f8 40%,#f6f7f8 100% );
  background-size: 80rem 14rem;
  -webkit-animation: placeholderShimmer 1s linear infinite forwards;
  animation: placeholderShimmer 1s linear infinite forwards;
}

.c1 img {
  display: block;
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.c3 {
  display: block;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: rgba(27,35,41,0.8);
}

.c3 > a {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  padding: 1.6rem;
  position: relative;
  height: 100%;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  color: #FFF;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4 {
  margin: 0;
  font-size: 2.0rem;
  margin-bottom: 0.8rem;
}

.c5 {
  color: #FFF;
  position: absolute;
  right: 1.6rem;
  top: 1.6rem;
  cursor: pointer;
}

.c5 svg {
  width: 2.5rem;
}

@media (min-width:768px) {
  .c3 {
    display: none;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <img
      alt="Pele: Birth of a Legend movie poster"
      src="https://m.media-amazon.com/images/M/MV5BMTg2NzQwMzQyMF5BMl5BanBnXkFtZTgwNzkzODk2ODE@._V1_SX300.jpg"
    />
  </div>
  <div
    class="c2 c3"
  >
    <a
      href="/movie/tt0995868"
    >
      <h2
        class="c4"
      >
        Pele: Birth of a Legend
      </h2>
      <span
        class=""
      >
        2016
      </span>
    </a>
    <div
      class="c5"
      role="button"
    >
      <svg
        aria-hidden="true"
        aria-label="Add to Favorites"
        class="c6"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
        />
      </svg>
    </div>
  </div>
</div>
`
